### PR TITLE
feat: add `#[borsh(crate = ...)]` item-level attribute

### LINKED ISSUE #95: Add derive macro attribute to specify borsh crate
Add derive macro attribute like [serde container's `crate` attr](https://serde.rs/container-attrs.html#crate) ([test](https://github.com/serde-rs/serde/blob/1fcda0ebdb2cfa050df42fded93ba07c14e4b683/test_suite/tests/test_serde_path.rs)):

1. To be able to work with byte blobs built with legacy borsh.
1. To be able to use re-exported borsh crate for `BorshSerialize`, `BorshDeserialize` derives. 

```rs
#[derive(borsh::BorshSerialize)]
#[borsh(crate = "legacy_borsh")]
struct NearLog;
```

### PR DESCRIPTION
**Pull request by dj8yfo**
resolves #95 
Need for this discovered in scope of work in [near-sdk-rs update](https://github.com/near/near-sdk-rs/pull/1075)

### CODE REVIEW

**Comment by dj8yfo** [comment]
Sample [crate](https://github.com/dj8yfo/workspace_with_a_borshik_reexport) with this attribute tested.

File: borsh-derive/src/internals/attributes/item/mod.rs [hunk 1]
```diff
@@ -85,17 +83,39 @@ pub(crate) fn contains_use_discriminant(input: &ItemEnum) -> Result<bool, syn::E
     Ok(use_discriminant.unwrap_or(false))
 }
 
-pub(crate) fn contains_initialize_with(attrs: &[Attribute]) -> Option<Path> {
+pub(crate) fn contains_initialize_with(attrs: &[Attribute]) -> Result<Option<Path>, Error> {
     let mut res = None;
     let attr = attrs.iter().find(|attr| attr.path() == BORSH);
     if let Some(attr) = attr {
-        let _ = attr.parse_nested_meta(|meta| {
```
Comment (dj8yfo) [inline borsh-derive/src/internals/attributes/item/mod.rs:92]: parsing error was silenced here, allowing to declare following value for the attribute without any `syn::Error` and returning `None`

```rust
#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
#[borsh(init={strange; blocky})]
struct A {
    lazy: Option<u64>,
}
```

File: borsh-derive/src/internals/attributes/item/mod.rs [hunk 2]
```diff
@@ -284,8 +303,8 @@ mod tests {
         }
             })
         .unwrap();
-        let actual = check_item_attributes(&item_struct);
-        local_insta_assert_snapshot!(actual.unwrap_err().to_token_stream().to_string());
+        let actual = check_attributes(&item_struct);
+        local_insta_assert_debug_snapshot!(actual.unwrap_err());
     }
     #[test]
     fn test_init_function() {
```
Comment (dj8yfo) [inline borsh-derive/src/internals/attributes/item/mod.rs:310]: this test doesn't change due to [impl-ToTokens-for-Option](https://docs.rs/quote/1.0.33/quote/trait.ToTokens.html#impl-ToTokens-for-Option%3CT%3E)

File: borsh-derive/src/internals/attributes/item/mod.rs [hunk 3]
```diff
@@ -85,29 +83,49 @@ pub(crate) fn contains_use_discriminant(input: &ItemEnum) -> Result<bool, syn::E
...
+    let attr = attrs.iter().find(|attr| attr.path() == BORSH);
+    if let Some(attr) = attr {
+        attr.parse_nested_meta(|meta| {
+            if meta.path == CRATE {
+                let value_expr: Path = parsing::parse_lit_into(BORSH, CRATE, &meta)?;
+                res = Some(value_expr);
+            }
+            if meta.path == USE_DISCRIMINANT || meta.path == INIT {
+                let _value_expr: Expr = meta.value()?.parse()?;
+            };
```
Comment (frol) [inline borsh-derive/src/internals/attributes/item/mod.rs:117]: Can you help me understand why we need to parse something completely unrelated here?

**Review by frol — APPROVED** [review]
Overall, looks great to me!

File: borsh-derive/src/internals/attributes/item/mod.rs (hunk 3 above)
Comment (dj8yfo) [reply borsh-derive/src/internals/attributes/item/mod.rs:117]: If we comment out these 115-117 highlighted lines, we get a failure in [test](https://github.com/near/borsh-rs/blob/crate_attribute/borsh-derive/src/internals/attributes/item/mod.rs#L389-L416) with following message:

```bash
---- internals::attributes::item::tests::test_init_function_with_use_discriminant_with_crate stdout ----
thread 'internals::attributes::item::tests::test_init_function_with_use_discriminant_with_crate' panicked at 'called `Result::unwrap()` on an `Err` value: Error("expected `,`")', borsh-derive/src/internals/attributes/item/mod.rs:412:20
```

The recommended parsing [helper](https://docs.rs/syn/latest/syn/struct.Attribute.html#method.parse_nested_meta)
parses the args `init = initialization_method, crate = "reexporter::borsh", use_discriminant=true` of `#[borsh(init = initialization_method, crate = "reexporter::borsh", use_discriminant=true)]` with following
[code](https://docs.rs/syn/latest/src/syn/meta.rs.html#384-399) somewhere down the way:

```rust
pub(crate) fn parse_nested_meta(
    input: [ParseStream](https://docs.rs/syn/latest/src/syn/parse.rs.html#223),
    mut logic: impl FnMut([ParseNestedMeta](https://docs.rs/syn/latest/src/syn/meta.rs.html#163-166)) -> Result<()>,
) -> Result<()> {
    loop {
        let path = input.call(parse_meta_path)?;
        logic(ParseNestedMeta { path, input })?;
        if input.is_empty() {
            return Ok(());
        }
        input.parse::<Token![,]>()?;
        if input.is_empty() {
            return Ok(());
        }
    }
}
```
It expects to completely consume input `ParseStream<'a>`, without skipping anything.
If we comment out ` let _value_expr: Expr = meta.value()?.parse()?;` it returns an error `input.parse::<Token![,]>()?;` on a new  loop iteration on a `=` punctuation, presumably originating [here](https://docs.rs/syn/latest/src/syn/token.rs.html#1006).

**Pull request merged**